### PR TITLE
Add search pagination token to python client

### PR DESCRIPTION
**Pull request by sueann**
## What changes are proposed in this pull request?
 
Adding support for search pagination tokens in the python client ``search_runs`` API. Implements the API changes and pagination in the ``RestStore``, but defers the implementation in ``FileStore`` and ``SQLAlchemyStore`` to later work.

With this implementation, any existing backend store should not break - the existing `search_runs` implementation will overwrite `AbstractStore`'s `search_runs`, and `_search_runs` will not be required (except by lint).
 
## How is this patch tested?
 
Unit tests
 
### Is this a user-facing change? 

- [ ] No. You can skip the rest of this section.
- [X] Yes. Give a description of this change to be included in the release notes for MLflow users.
 
<TBD after RFC>
 
### What component(s) does this PR affect?
 
- [X] API 
- [X] Tracking
- [X] Python

### How should the PR be classified in the release notes? Choose one:
 
- [X] `rn/feature` - A new user-facing feature worth mentioning in the release notes

### CODE REVIEW

File: mlflow/store/abstract_store.py [hunk 1]
```diff
@@ -4,6 +4,13 @@
 from mlflow.store import SEARCH_MAX_RESULTS_DEFAULT
 
 
+class ListWithToken(list):
+
+    def __init__(self, items, token):
```
[the remark this answers is not in the document]
Comment (sueann) [reply mlflow/store/abstract_store.py]: hm... but `items` here could actually be any iterable, not just a list. and `items` would refer to the items in any such iterable. not sure if there is a better term for it?

File: mlflow/store/file_store.py [hunk 2]
```diff
@@ -545,7 +544,9 @@ def search_runs(self, experiment_ids, search_filter, run_view_type,
             run_infos = self._list_run_infos(experiment_id, run_view_type)
             runs.extend(self.get_run(r.run_id) for r in run_infos)
         filtered = [run for run in runs if not search_filter or search_filter.filter(run)]
-        return sorted(filtered, key=lambda r: (-r.info.start_time, r.info.run_id))[:max_results]
+        runs = sorted(filtered, key=lambda r: (-r.info.start_time, r.info.run_id))[:max_results]
+        token = "PAGINATION_TOKEN_NOT_IMPLEMENTED"
```
[the remark this answers is not in the document]
Comment (sueann) [reply mlflow/store/file_store.py]: makes sense, though from the UX perspective, would it be better to return `None` so it is easier to check for an invalid token? or should we handle such a token inside `search_runs` - I assume we will have to validate the token anyway - so the user doesn't have to worry about it.

File: mlflow/store/abstract_store.py [hunk 3]
```diff
@@ -209,7 +215,28 @@ def search_runs(self, experiment_ids, search_filter, run_view_type,
...
+        return ListWithToken(runs, token)
+
+    @abstractmethod
+    def _search_runs(self, experiment_ids, search_filter, run_view_type, max_results):
+        """
+        Return runs that match the given list of search expressions within the experiments.
+        Given multiple search expressions, all these expressions are ANDed together for search.
+
+        :param experiment_ids: List of experiment ids to scope the search
+        :param search_filter: :py:class`mlflow.utils.search_utils.SearchFilter` object to encode
```
[the remark this answers is not in the document]
Comment (sueann) [reply mlflow/store/abstract_store.py]: ah thanks for the pointer. i will need to do some conflict merging once https://github.com/mlflow/mlflow/pull/1444 is in.

**Comment by sueann** [comment]
Current status:
- Resolved merge conflicts with #1437
- Need to add the `token` parameter to the `search_runs` method
- Need to implement pagination for `FileStore` and `SQLAlchemyStore`

**Comment by sueann** [comment]
Will do the TODOs above in follow-up PRs. Would love to get this in soon to do that. Thanks!

File: mlflow/store/abstract_store.py [hunk 4]
```diff
@@ -195,20 +202,46 @@ def get_metric_history(self, run_id, metric_key):
         """
         pass
 
-    @abstractmethod
     def search_runs(self, experiment_ids, filter_string, run_view_type,
-                    max_results=SEARCH_MAX_RESULTS_DEFAULT, order_by=None):
+                    max_results=SEARCH_MAX_RESULTS_DEFAULT, order_by=None, page_token=None):
```
Comment (sueann) [inline mlflow/store/abstract_store.py]: @mparkhe is `None` an okay default value for the page token (based on how you plan to implement validation/handling of this argument in the server)?

Comment (sueann) [reply mlflow/store/abstract_store.py]: Actually looks like proto won't send the field if it's `None` so should be fine

File: mlflow/store/abstract_store.py [hunk 5]
```diff
@@ -4,6 +4,13 @@
 from mlflow.store import SEARCH_MAX_RESULTS_DEFAULT
 
 
+class PagedList(list):
```
Comment (mparkhe) [inline mlflow/store/abstract_store.py]: I like that name!

File: mlflow/store/abstract_store.py (hunk 4 above)
Comment (mparkhe) [reply mlflow/store/abstract_store.py]: 👍

File: mlflow/store/abstract_store.py [hunk 6]
```diff
@@ -195,20 +202,46 @@ def get_metric_history(self, run_id, metric_key):
...
+                     page_token):
+        """
+        Return runs that match the given list of search expressions within the experiments, as
+        well as a pagination token (indicating where the next page should start). Subclasses of
+        ``AbstractStore`` should implement this method to support pagination instead of
+        ``search_runs``.
+
+        See ``search_runs`` for parameter descriptions.
+
+        :param page_token:
```
Comment (mparkhe) [inline mlflow/store/abstract_store.py]: something is missing here.

File: mlflow/store/file_store.py [hunk 7]
```diff
@@ -533,8 +533,11 @@ def _list_run_infos(self, experiment_id, view_type):
                                 exc_info=True)
         return run_infos
 
-    def search_runs(self, experiment_ids, filter_string, run_view_type,
-                    max_results=SEARCH_MAX_RESULTS_THRESHOLD, order_by=None):
+    def _search_runs(self, experiment_ids, filter_string, run_view_type, max_results, order_by,
+                     page_token):
+        if page_token:
+            raise MlflowException("SQLAlchemy-backed tracking stores do not yet support pagination"
```
Comment (mparkhe) [inline mlflow/store/file_store.py]: `SQLAlchemy-backed tracking stores do` -> `FileStore does`

**Review by mparkhe — COMMENTED** [review]
small *nit*s but overall looks good.

**Review by mparkhe — APPROVED** [review]

**Pull request merged**